src/FrameBase.re

```diff
@@ -60,30 +60,65 @@ external waitForXPath :
   Js.Promise.t(ElementHandle.t) =
   "";
 
-/* TODO: Add support [, ...args] */
+type jsonOrHandle =
+  | JSON(serializable)
+  | Handle(JSHandle.t);
+
+/**
+ * selectOneEval(selector, fn, args, page)
+ * Runs document.querySelector in the page and passes it as the first
+ * argument to `fn`. Additional arguments to the `fn` can be provides in
+ * `args`. If there's no element matching selector, the method throws an
+ * error. If the function returns a Promise, it waits for the promise to
+ * resolve and returns its value.
+ */
 [@bs.send.pipe : t]
-external selectOneEval : (string, unit => unit) => Js.Promise.t('a) = "$eval";
-
-/* TODO: Add support [, ...args] */
+[@bs.splice]
+external selectOneEval :
+  (string, 'a => 'b, array(jsonOrHandle)) => Js.Promise.t(serializable) =
+  "$eval";
+
+/**
+ * selectAllEval(selector, fn)
+ * Runs document.querySelectorAll in the page and passes it as the first
+ * argument to `fn`. If `fn` returns a Promise, then it waits for
+ * the promise to resolve and returns its value.
+ */
 [@bs.send.pipe : t]
-external selectAllEval : (string, unit => unit) => Js.Promise.t('a) = "$$eval";
+external selectAllEval :
+  (string, Dom.element => 'a) => Js.Promise.t(serializable) =
+  "$$eval";
+
+[@bs.send.pipe : t] [@bs.splice]
+external selectAllEvalWithArgs :
+  (string, Dom.element => 'a, array(jsonOrHandle)) =>
+  Js.Promise.t(serializable) =
+  "$$eval";
 
 [@bs.send.pipe : t]
 external addScriptTag : tagOptions => Js.Promise.t(ElementHandle.t) = "";
 
 [@bs.send.pipe : t]
 external addStyleTag : tagOptions => Js.Promise.t(ElementHandle.t) = "";
 
-/* TODO: Currently only ever work for functions taking no arguments,
-   and the second parameter array can only ever be empty */
 [@bs.send.pipe : t] [@bs.splice]
 external evaluate :
-  (unit => Js.Promise.t(Js.Json.t), array({..})) => Js.Promise.t(Js.Json.t) =
+  (Eval.Fn.t, array(Eval.Arg.t)) => Js.Promise.t(serializable) =
```

Comment:
Since I wound up splitting `evaluate()` into `evaluateString` and this function, we could maybe rethink whether we want to use `Eval.Fn.t` as the first argument, or if `Eval.Fn.t` needs to include `string`. We need to somehow indicate how many args the function takes, or require the user to curry all the args before passing the function to `evaluate`.